controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1420,6 +1485,87 @@ private TxnStatus handleDataNotFoundException(Throwable ex) {
+        // if time is not advanced --> respond with invalid time.
+        // if position is not advanced --> respond with invalid position
+        ImmutableMap<Long, Long> newPosition = ImmutableMap.copyOf(position);
+        return Futures.exceptionallyExpecting(getWriterMarkRecord(writer), DATA_NOT_FOUND_PREDICATE, null) 
+                .thenCompose(record -> {
+                    if (record == null) {
+                        // Attempt to create a new record. It is possible that while we are attempting to create the record, 
+                        // a concurrent request could have created it. 
+                        // For example: a writer sends a request to one of controller instances to note writer mark and 
+                        // the connection is dropped. It could then send the request to noteTime to a different controller 
+                        // instance. So two controller instances could attempt concurrent creation of mark for the said writer.
+                        // In this case, we may have found mark to be inexistent when we did a get but it may have been created 
+                        // when we attempt to create it. So immediately after attempting a create, if we get DataExists, 
+                        // we will call noteWriterMark recursively. 
+                        return Futures.exceptionallyComposeExpecting(
+                                createWriterMarkRecord(writer, timestamp, newPosition)
+                                        .thenApply(v -> WriterTimestampResponse.SUCCESS),
+                                e -> Exceptions.unwrap(e) instanceof StoreException.DataExistsException, 
+                                () -> noteWriterMark(writer, timestamp, position));
+                    } else {
+                        // sanity check and update
+                        if (record.getObject().getTimestamp() > timestamp) {
+                            // existing time is already ahead of new time
+                            return CompletableFuture.completedFuture(WriterTimestampResponse.INVALID_TIME);
+                        }
+                        if (!compareWriterPositions(record.getObject().getPosition(), newPosition)) {
+                            // existing position is already ahead of new position
+                            return CompletableFuture.completedFuture(WriterTimestampResponse.INVALID_POSITION);
+                        }
+                        
+                        // its a valid mark, update it
+                        return updateWriterMarkRecord(writer, timestamp, newPosition, record.getVersion())
+                                .thenApply(v -> WriterTimestampResponse.SUCCESS);
+                    }
+                });
+    }
+
+
+    /**
+     * Compares two given positions and returns true if position 2 is ahead of position 1. 
+     * 
+     * Note: This is not absolutely correct implementation because that would be a very costly computation. 
```

Comment:
Typically a writer reporting a mark will always advance time as well with each new mark. 

And if this method is called from commit-transaction, then the commit happens against a specific epoch. So the highest segment number check suffices there. 

And as long as writer is advancing time in its transaction commits as well, the check to compare writer marks is more of a sanity check to prevent rouge writers or tackle some bug.